versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/MongoDbStore.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.dremio.nessie.versioned.store.mongodb;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.dremio.nessie.versioned.ReferenceNotFoundException;
+import com.dremio.nessie.versioned.impl.InternalRef;
+import com.dremio.nessie.versioned.impl.condition.ConditionExpression;
+import com.dremio.nessie.versioned.impl.condition.UpdateExpression;
+import com.dremio.nessie.versioned.store.Id;
+import com.dremio.nessie.versioned.store.LoadOp;
+import com.dremio.nessie.versioned.store.LoadStep;
+import com.dremio.nessie.versioned.store.SaveOp;
+import com.dremio.nessie.versioned.store.Store;
+import com.dremio.nessie.versioned.store.ValueType;
+import com.google.common.collect.ListMultimap;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoDatabase;
+
+/**
+ * This class implements the Store interface that is used by Nessie as a backing store for versioning of it's
+ * Git like behaviour.
+ * The MongoDbStore connects to an external MongoDB server.
+ */
+public class MongoDbStore implements Store {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbStore.class);
+
+  private final String connectionString;
+  private final String databaseName;
+
+  // The client that connects to the MongoDB server.
+  protected MongoClient mongoClient;
+  // The database hosted by the MongoDB server.
+  private MongoDatabase mongoDatabase;
+
+  /**
+   * Creates a store ready for connection to a MongoDB instance.
+   * @param connectionString the URL of the database with which to connect. Eg mongodb://localhost:27017.
+   * @param databaseName the name of the database to retrieve
+   */
+  public MongoDbStore(String connectionString, String databaseName) {
+    this.connectionString = connectionString;
+    this.databaseName = databaseName;
+  }
+
+  /**
+   * Creates a connection to an existing database or creates the database if it does not exist.
```

Comment:
I don't believe this is totally accurate, GetDatabase doesn't really create a Database, it only returns a MongoDatabase representing the specified database.
The database is lazily created when an operation like a create collection is called.
http://mongodb.github.io/mongo-java-driver/3.7/javadoc/com/mongodb/MongoClient.html#getDatabase-java.lang.String-

Reply:
Fair point. I will update the wording.

Reply:
Updated wording.